### PR TITLE
drivers: flash: stm32g4: Fix page erase modulo 127

### PR DESCRIPTION
**Pull request by alexanderwachter**
The stm32g4 flash is split into two banks in the default configuration.
When trying to erase memory in the upper half of the memory, the page number was taken modulo 127 due to a seven-bit register. This causes an erase at the wrong address.

### CODE REVIEW

**Comment by alexanderwachter** [comment]
Ping @rosterloh

**Comment by rosterloh** [comment]
The device I tested this on is the STM32G431RB which only has 128K of flash. Which device are you using @alexanderwachter ?

**Comment by alexanderwachter** [comment]
g474re, which has 512k flash.

**Review by rosterloh — APPROVED** [review]
LGTM

**Review by FRASTM — APPROVED** [review]

File: drivers/flash/flash_stm32g4x.c [hunk 1]
```diff
@@ -99,10 +102,26 @@ static int erase_page(struct device *dev, unsigned int page)
 		return rc;
 	}
 
+	if (page > 127) {
```
Comment (erwango) [inline drivers/flash/flash_stm32g4x.c]: Please put this under `#ifdef FLASH_OPTR_DBANK`

File: drivers/flash/flash_stm32g4x.c [hunk 2]
```diff
@@ -113,7 +132,7 @@ static int erase_page(struct device *dev, unsigned int page)
 	/* Wait for the BSY bit */
 	rc = flash_stm32_wait_flash_idle(dev);
 
-	regs->CR &= ~FLASH_CR_PER;
```
Comment (erwango) [inline drivers/flash/flash_stm32g4x.c]: Please put this under #ifdef FLASH_OPTR_DBANK

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for having added the LOG prints.
2 comments, otherwise this is fine.

File: drivers/flash/flash_stm32g4x.c (hunk 1 above)
Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c]: The register has only 6 bits. Numbers > 127 are never allowed.

File: drivers/flash/flash_stm32g4x.c (hunk 2 above)
Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c]: It does not change anything. The datasheet tells to keep FLASH_CR_BKER at reset value (false).

Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c]: Oh, I see. Some MCUs don't have that bit!

File: drivers/flash/flash_stm32g4x.c (hunk 1 above)
Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c]: Done

File: drivers/flash/flash_stm32g4x.c (hunk 2 above)
Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c]: Done

File: drivers/flash/flash_stm32g4x.c [hunk 3]
```diff
@@ -99,10 +102,29 @@ static int erase_page(struct device *dev, unsigned int page)
...
+	}
+
+
+	__ASSERT(page <= 127, "There are only 127 pages, but page is %d", page);
+#endif
+
 	/* Set the PER bit and select the page you wish to erase */
 	regs->CR |= FLASH_CR_PER;
 	regs->CR &= ~FLASH_CR_PNB_Msk;
-	regs->CR |= ((page % 256) << 3);
```
Comment (alexanderwachter) [inline drivers/flash/flash_stm32g4x.c]: @erwango  I removed the %256. Is that allowed for the other SOCs as well?

Comment (erwango) [reply drivers/flash/flash_stm32g4x.c]: Yes, this one is available on all G4 variants.
So this line is equivalent ?

Comment (erwango) [reply drivers/flash/flash_stm32g4x.c]: Are you sure this is correct ?

Comment (erwango) [reply drivers/flash/flash_stm32g4x.c]: ok, right

**Review by erwango — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**